Add flash attention forward op

harsh-nod opened the pull request:
This patch adds the flash attention forward op
to LinalgExt. For tiling, we expose the two
outer dimensions corresponding to batch and sequence length.

harsh-nod committed:
Add flash attention forward op

This patch adds the flash attention forward op
to LinalgExt. For tiling, we expose the two
outer dimensions corresponding to batch and sequence
length.

harsh-nod committed:
Add lowering of flash attention to linalg ops

This patch adds a pass to lower the flash attention
forward op to a tiled version of 2 matmuls + softmax
that can then be run on a particular backend.

harsh-nod committed:
Attach flash attention op to interfaces

Also fix bugs and update corresponding unit tests.

ftynse reviewed (commented):
What Nicolas means by "functional-style" is expressing the top-level rewrite rule as a function that takes a `FlashAttnForwardOp` and returns the most significant ops the rewrite produced. This can also be done in a pattern, see https://github.com/llvm/llvm-project/blob/ebfb1ddbe18457723ea62f4ecc94667a44ca4569/mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp#L680 for example. Thus, one can take the result of the rewrite and directly apply another rewrite rule on it, without going over the IR and matching or the rewriter driver. If we were in a sufficiently advanced functional language, we would have been able to do a function composition of the two rewrite rules (provided they return their most significant ops) to obtain a single rewrite rule.

MaheshRavishankar reviewed (changes requested):
Sorry for the delay in review. Mostly looks fine. But a big part of the `TileAndDecompose...` is hard to read with some more comments. Ill do another round shortly.

harsh-nod committed:
Address Alex, Nicolas and Mahesh's comments

Also fix a bug that was updating the wrong
iter args. This now lowers through the
CUDA backend.

harsh-nod committed:
This patch fixes issues with functional correctness.

The patch now outputs the correct result when
using the CPU backend.
